oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,450 @@
+edX learners, educators, researchers, and Open edX operators. Those are out of scope of this OEP and maintained
+by the edX documentation team at docs.edx.org_.
+
+Developer documentation is inconsistent and poorly maintained. Documents are also difficult to find. They
+exist in multiple locations, including `Read the Docs`_, GitHub_ repositories, Confluence_ wiki, and `Google Docs`_.
+
+This proposal focuses on long-term developer *Documentation* needs. It does not concern itself with any transitory
+*Notes* that a development team may use for project management, brainstorming, meeting notes, etc. Code *Comments*
+that a developer may add to clarify or rationalize a code segment are also out of scope of this OEP.
+
+.. _Read the Docs: https://readthedocs.org
+.. _Confluence: https://openedx.atlassian.net/wiki
+.. _GitHub: https://github.com/edx
+.. _Google Docs: https://docs.google.com
+
+
+Decisions
+*********
+
+.. image:: oep-0019/developer_docs.png
+   :alt: A diagram that shows the different types of documentation and their
+    relationships. This is further described below.
+
+Agile Documentation Types
+=========================
+
+With Agile development and `Clean Code`_, we can minimize documentation (and its maintainence) since code should
+be readable and self-documenting. However, for long-term maintainence of a platform, the following types of developer
+documentation are still needed:
+
+.. _Clean Code: https://www.oreilly.com/library/view/clean-code/9780136083238/
+
+.. contents::
+   :local:
+   :depth: 1
+
+OEPs
+====
+
+.. list-table::
+   :widths: 15 75 
+
+   * - **Description**
+     - Open edX Proposals (OEPs) are a collection of system-wide technology decisions made by the Open edX
+       community, in the form of best practices, architecture designs, and processes.
+   * - **Answers the questions**
+     - 
+       * *What are the best practices and processes for developing on the Open edX platform?*
+       * *How does system-wide framework X work on the Open edX platform? (e.g., Analytics, Authorization, etc).*
+       * *Why was system-wide technology X or design Y chosen for the Open edX platform?*
+   * - **Rationale**
+     - To establish software development coherency and efficiency with minimal centralized governance
+       across a long-term maintained platform.
+   * - **Location in GitHub**
+     - Stored and reviewed in the `open-edx-proposals`_ GitHub repo; published to `Open edX Proposals Read The Docs`_.
+   * - **Maintainence**
+     - OEP-1_ describes the OEP review process. They will be maintained by the `edX Architecture team`_.
+
+.. _open-edx-proposals: https://github.com/edx/open-edx-proposals
+.. _Open edX Proposals Read The Docs: https://open-edx-proposals.readthedocs.io
+.. _OEP-1: https://open-edx-proposals.readthedocs.io/en/latest/oep-0001.html
+.. _edX Architecture team: https://openedx.atlassian.net/wiki/spaces/AC/pages/439353453/Architecture+Team
+
+ADRs
```

Comment:
@mikedikan Thanks for asking.  🎤

As of now, here is what we have been thinking:

- If the decision is a system-wide decision, then it should be recorded in an OEP ([see L84](https://github.com/edx/open-edx-proposals/pull/74/files#diff-63e65fcef1f4cd3676f373d1bea15029R84)).

- If the decision affects multiple repos but isn't "system-wide" or doesn't necessarily need to go through an OEP process, it can be recorded in a repo that makes sense for that part of the system.  For example, it can be in a top-level repo of a set of dependent repos or it can be in a new docs repo created for that subsystem.  This form of decentralization of decision docs is necessary as the system scales.  However, at this time, our platform is still small enough that creating an OEP is most likely a better choice.

- Finally, note that we have a system-wide developer docs repo, edx-developer-docs, that can also be used to contain cross-repo documentation ([see L243](https://github.com/edx/open-edx-proposals/pull/74/files#diff-63e65fcef1f4cd3676f373d1bea15029R243)).

Reply:
I don't know what the difference is between 'system-wide' and 'affects multiple repos'.  Does system-wide mean every repo is affected?  A majority?

Reply:
"System-wide" is not dependent on the number of repos affected.  It is a matter of the overall impact on the entire codebase and whether the decision should follow the OEP review process.  For example, if it affects multiple teams, then it should be an OEP.  If it provides a basic framework for others to build upon, then it should be an OEP.  We imagine this will become clearer as we create more ADRs and OEPs.  Please ask in the #architecture slack channel on a case-by-case basis if/when you have questions.  We have [updated the OEP](42da7bbbe12e60f08d7262b1f939bfbdac634024) to add a note that some things will still be figured out as we go.